File: src/trading/jbtc/parser.ts
//? This script parses the JBTC value list excel
//? DO NOT RUN THE PARSER SCRIPT IF YOU HAVENT INITIALIZED GOOGLE API KEY INSIDE "secrets.json"

import { google } from "googleapis";
import { JBItemDemand, StrToJBDemand, type JBItem } from "../types";
import { writeFileSync } from "node:fs"

const final: JBItem[] = [
    {
        name: "Volt Bike",
        value: 1_000_000,
        demand: JBItemDemand.Mid
    },
    {
        name: "Blackhawk",
        value: 1_000_000,
        demand: JBItemDemand.Mid
    },
    {
        name: "Trailblazer",
        value: 1_000_000,
        demand: JBItemDemand.Mid
    },
    {
        name: "Monster",
        value: 1_000_000,
        demand: JBItemDemand.Low
    },
    {
        name: "Jet",
        value: 1_000_000,
        demand: JBItemDemand.Mid
    },
    {
        name: "Drone",
        value: 1_000_000,
        demand: JBItemDemand.Mid
    }
];

type Response = string[]

async function parse(sheets: any, range: string, hyper?: boolean, format?: (id: string) => string) {
    const result = await sheets.spreadsheets.values.get({
        spreadsheetId: "12aPBmrHP5MLwoht9QcBiERPYUmXTJiTbBF43630togE",
        range: range
    });
    const values = result.data.values;
    values?.forEach(([id, value, duped_value, demand, notes]: Response) => {
        // console.log(id, value, duped_value, demand, notes);
        if (hyper && format) {
            final.push({
                name: id.trim() !== "Hypershift" ? format(id) : "Hypershift",
                value: Number(value.replaceAll(",", "")),
                // duped_value: Number(duped_value.replaceAll(",", "")),
                demand: StrToJBDemand(duped_value),
                notes: demand?.trim().length == 0 ? undefined : demand
            });
        } else {
            final.push({
                name: id,
                value: Number(value.replaceAll(",", "")),
                duped_value: Number(duped_value.replaceAll(",", "")),
                demand: StrToJBDemand(demand),
                notes: notes?.trim().length == 0 ? undefined : notes
            });
        }
    })
    console.log("[JBTC] Completed", range);
}

async function main() {
    const auth = await google.auth.getClient({
        scopes: ["https://www.googleapis.com/auth/spreadsheets.readonly"],
    });
    const sheets = google.sheets({ version: "v4", auth });
    await parse(sheets, "Value List!C20:G67"); // Vehicles
    await parse(sheets, "Value List!C71:G120") // Txts/Colors
    await parse(sheets, "Value List!C124:G183") // Rims
    await parse(sheets, "Value List!C187:G245") // Spoilers
    await parse(sheets, "Value List!C249:G267") // Tires/Horns
    await parse(sheets, "Value List!C249:G267") // Furniture
    await parse(sheets, "Value List!C249:G267") // Gun Skins

    // Hypers
    await parse(sheets, "Hyperchromes!C22:F30", true, (n) => `Hyper ${n} L5`)
    await parse(sheets, "Hyperchromes!C34:F41", true, (n) => `Hyper ${n} L4`)
    await parse(sheets, "Hyperchromes!C45:E52", true, (n) => `Hyper ${n} L3`)
    await parse(sheets, "Hyperchromes!C56:E63", true, (n) => `Hyper ${n} L2`)
    // No hyper lvl 1s!!
    writeFileSync("./src/trading/jbtc/values.json", JSON.stringify(final, null, 2));
    console.log("[JBTC] Saved");
}

main();

File: src/jailbreak/models/jbtc/parser.rs
use std::fs;

use reqwest::blocking as breqwest;
use serde_json::Value;

use crate::jailbreak::traits::{self, ReplaceAll};
use crate::jailbreak::types::{JBItem, JBItemDemand, JBItemType, JBTerminalConfig};
// TODO

const SPREADSHEET_ID: &str = "12aPBmrHP5MLwoht9QcBiERPYUmXTJiTbBF43630togE";

fn get_spreadsheet_data(range: &str) -> Result<Vec<Value>, reqwest::Error> {
    let api_key = JBTerminalConfig::new().api_key;
    let url = format!(
        "https://sheets.googleapis.com/v4/spreadsheets/{}/values/{}?key={}",
        SPREADSHEET_ID, range, api_key
    );

    let response = breqwest::get(&url)?;
    let body = response.text()?;
    let data: serde_json::Value = serde_json::from_str(&body).unwrap();
    let cells = data.get("values").unwrap().as_array().unwrap();
    Ok(cells.clone())
}

pub fn parse_values(
    range: &str,
    values: &mut Vec<JBItem>,
    mut category: JBItemType,
    str_add: &str,
) {
    let cells = get_spreadsheet_data(range).unwrap();
    for cell in cells {
        let cell = cell.as_array().unwrap();
        let mut name = cell.get(0).unwrap().as_str().clone().unwrap().to_string();
        let value = cell
            .get(1)
            .unwrap()
            .as_str()
            .clone()
            .unwrap()
            .to_string()
            .replace_all(",", "")
            .parse::<u32>()
            .unwrap();
        let mut duped_value = None;
        let mut demand = JBItemDemand::None;
        let mut notes = None;
        if category != JBItemType::Hyperchrome {
            duped_value = cell
                .get(2)
                .unwrap()
                .as_str()
                .clone()
                .unwrap()
                .to_string()
                .replace_all(",", "")
                .parse::<u32>()
                .ok();
            demand = JBItemDemand::from(cell.get(3).unwrap().as_str().unwrap().to_string());
            if let Some(value) = cell.get(4) {
                notes = Some(value.as_str().unwrap().to_string());
            }
        } else {
            name = "Hyper ".to_owned() + &name;
            demand = JBItemDemand::from(cell.get(2).unwrap().as_str().unwrap().to_string());
            if let Some(value) = cell.get(3) {
                notes = Some(value.as_str().unwrap().to_string());
            }
        }
        // TODO notes be optional not ""

        if name.to_lowercase().contains("horn") {
            category = JBItemType::VehicleHorn;
        } else if name.to_lowercase().contains("drift") {
            category = JBItemType::DriftParticle;
        } else if name.to_lowercase().contains("radiant")
            || name.as_str() == "Gold"
            || name.as_str() == "Fall chrome"
        {
            category = JBItemType::Color;
        }

        // println!("{name} {value} {duped_value:?} {notes}");
        name.push_str(str_add);
        values.push(JBItem {
            name,
            value,
            demand,
            duped_value,
            notes,
            og: None,
            category,
        })
    }
}

pub fn parse() -> Result<(), reqwest::Error> {
    let api_key = JBTerminalConfig::new().api_key;
    if api_key.is_empty() {
        return Ok(()); // Not ok but idc
    }
    let mut values: Vec<JBItem> = Vec::new();
    parse_values("Value List!C20:G67", &mut values, JBItemType::Vehicle, "");
    parse_values("Value List!C71:G120", &mut values, JBItemType::Texture, "");
    parse_values("Value List!C124:G183", &mut values, JBItemType::Rim, "");
    parse_values("Value List!C187:G245", &mut values, JBItemType::Spoiler, "");
    parse_values("Value List!C249:G267", &mut values, JBItemType::Tires, "");
    parse_values(
        "Value List!C271:G308",
        &mut values,
        JBItemType::Furniture,
        "",
    );
    parse_values(
        "Value List!C312:G321",
        &mut values,
        JBItemType::GunTexture,
        "",
    );

    // Hyperchromes
    parse_values(
        "Hyperchromes!C22:F30",
        &mut values,
        JBItemType::Hyperchrome,
        " level V",
    );
    parse_values(
        "Hyperchromes!C34:F41",
        &mut values,
        JBItemType::Hyperchrome,
        " level IV",
    );
    parse_values(
        "Hyperchromes!C45:F52",
        &mut values,
        JBItemType::Hyperchrome,
        " level III",
    );
    parse_values(
        "Hyperchromes!C56:F63",
        &mut values,
        JBItemType::Hyperchrome,
        " level II",
    );
    // No hyperchromes for lvl 1

    fs::write("data/jbtc.json", serde_json::to_string(&values).unwrap()).unwrap();

    Ok(())
}
